tornado/test/web_test.py

```diff
@@ -1404,6 +1404,30 @@ def test_root_static_path(self):
         response = self.get_and_head("/root_static" + urllib.parse.quote(path))
         self.assertEqual(response.code, 200)
 
+    def test_same_directory_traversal(self):
+        response = self.fetch("/static/./password.txt", follow_redirects=False)
+        self.assertEqual(response.code, 301)
+        self.assertTrue(response.headers["Location"].endswith("/static/password.txt"))
+
+    def test_inside_directory_traversal(self):
+        response = self.fetch("/static/../static/password.txt", follow_redirects=False)
+        self.assertEqual(response.code, 301)
+        self.assertTrue(response.headers["Location"].endswith("/static/password.txt"))
+
+    def test_uri_encoding(self):
+        response = self.fetch("/static/password.tx%74", follow_redirects=False)
+        self.assertEqual(response.code, 301)
+        self.assertTrue(response.headers["Location"].endswith("/static/password.txt"))
+
+    def test_double_uri_encoding(self):
+        response = self.fetch("/static/password.tx%2574", follow_redirects=False)
+        self.assertEqual(response.code, 301)
+        self.assertTrue(response.headers["Location"].endswith("/static/password.tx%74"))
```

Comment:
I don't think this is the right result. It will redirect to `password.txt` per the previous url, but `password.txt` and `password%2574.txt` are two distinct urls (the latter should serve the file named `password%74.txt` if it exists, and otherwise be a 404)

Reply:
If you want to support `%` in filenames AND opening `password.tx%74` as `password.txt`, this cannot be solved via redirect.